src/client/src/producer.rs

```diff
@@ -61,10 +61,19 @@ impl TopicProducer {
     )]
     pub async fn send<K, V>(&self, key: K, value: V) -> Result<(), FluvioError>
     where
-        K: Into<Vec<u8>>,
-        V: Into<Vec<u8>>,
+        K: Into<RecordKey>,
+        V: Into<RecordValue>,
     {
-        self.send_all(Some((Some(key.into()), value))).await?;
+        let record_key = key.into();
```

Comment:
Instead of just modifying this.  We should update `send_all` to use `RecordKey` and `RecordValue`.

Reply:
Hmm I don't think we can actually do this. If you look at `send_all`:

```rust
    pub async fn send_all<K, V, I>(&self, records: I) -> Result<(), FluvioError>
    where
        K: Into<Vec<u8>>,
        V: Into<Vec<u8>>,
        I: IntoIterator<Item = (Option<K>, V)>,
    { ... }
```

Specifically this part: `I: IntoIterator<Item = (Option<K>, V)>`

This means that `send_all` is using Option::None to represent "no key" already. If we were to try to change it like this:

```rust
    pub async fn send_all<K, V, I>(&self, records: I) -> Result<(), FluvioError>
    where
        K: Into<RecordKey>,
        V: Into<RecordValue>,
        I: IntoIterator<Item = (Option<K>, V)>,
    { ... }
```

Then we would have an odd set of possible inputs for the key items:

```rust
let records = vec![ (None, "World") ];
let records = vec![ (Some(RecordKey::Null), "World") ]; // This means the same as above
let records = vec![ ("Hello", "World") ];
```

I think this would make things confusing since there would be more than one correct way to do things, which is good to avoid. Because of this, I don't think we should change the signature of `send_all`.

Reply:
This works:
```
use super::*;

    use dataplane::bytes::Bytes;

    pub enum RecordKey {
        Null,
        Key(Bytes),
    }
    
    impl<K: Into<Vec<u8>>> From<K> for RecordKey 
    {
        fn from(key: K) -> Self {
            let vec: Vec<u8> = key.into();
            Self::Key(vec.into())
        }
    }
    
    pub struct RecordValue(Bytes);
    
    impl<T: Into<Vec<u8>>> From<T> for RecordValue {
        fn from(val: T) -> Self {
            let vec: Vec<u8> = val.into();
            Self(vec.into())
        }
    }

    pub async fn send_all<K, V, I>(records: I) -> Result<(), FluvioError>
    where
        K: Into<RecordKey>,
        V: Into<RecordValue>,
        I: IntoIterator<Item = (K,V)>,
    { 
        Ok(())
     }


    async fn test() {
        send_all(vec![ (RecordKey::Null, "World") ]).await;
        send_all(vec![ ("Hello", "World")]).await;
    }
```

Reply:
The problem with this solution is that it would be a breaking change. Changing `K: Into<Vec<u8>>` to `K: Into<RecordKey>` is fine because of the `impl<K: Into<Vec<u8>>> From<K> for RecordKey`, but the breaking change comes from this change:

- `I: IntoIter<Item = (Option<K>, V)>` into
- `I: IntoIter<Item = (K, V)>`

So if we wanted to adopt that change to the API, it would be a breaking change.

Reply:
Since this is new API, we can make breaking changes assuming next release can be done quickly.  I think it is worth it, this will reduce API surface and make it consistent with `send`

Reply:
Ok, but I would still like to bump the fluvio crate from `0.7.0` to `0.8.0` so that cargo cannot accidentally confuse the two